### PR TITLE
mgr/cephadm: add ok-to-stop functions for ceph client services

### PR DESCRIPTION
**Pull request by Daniel-Pivonka**
create ok-to-stop functions for client services.

depends on #38854 
related to #38897 

Signed-off-by: Daniel-Pivonka <dpivonka@redhat.com>
Signed-off-by: Adam King <adking@redhat.com>

### CODE REVIEW

**Comment by Daniel-Pivonka** [comment]
created as a draft because we are still testing but wanted to get something up to share and get feedback from

**Review by mgfritch — COMMENTED** [review]
We appear to duplicate a fair amount of logic to check if the num of dameons >=1. Can this be consolidated up into `ok_to_stop` of the CephadmService class?

File: src/pybind/mgr/cephadm/services/cephadmservice.py [hunk 1]
```diff
@@ -737,6 +738,32 @@ def create_zone() -> None:
...
             self.mgr.log.info('updated period')
 
+    def ok_to_stop(self, daemon_ids: List[str], force: bool = False) -> HandleCommandResult:
+        # if load balancer (ha-rgw) is present block if only 1 daemon up otherwise ok
+        # if no load balancer, warn if > 1 daemon, block if only 1 daemon
+        def ha_rgw_present() -> bool:
+            for name, spec in self.mgr.spec_store.specs.items():
+                if spec.service_type == 'ha-rgw':
+                    return True
+            return False
```
[the remark this answers is not in the document]
Comment (jmolmo) [reply src/pybind/mgr/cephadm/services/cephadmservice.py:748]: @Daniel-Pivonka : probably we can consider that `ha_proxy` is running properly verifying the `size` and `running` attributes of the Service spec. if "size == running" then we can infer that we have ha-rgw deployed and running properly.
You can see an example of [use of this two attributes](https://github.com/ceph/ceph/blob/aab9de639f890fd05667d33a15afc66bc1591409/src/pybind/mgr/orchestrator/module.py#L593) in the "orch ls" command:

What do you think?

**Review by jmolmo — CHANGES_REQUESTED** [review]
I think this is what we need!.. Just to say that it could be nice to improve the detection of ha-rwg running properly with my suggestion.

**Comment by adk3798** [comment]
Removed ha-rgw ok-to-stop from this. This PR is meant to be more straightforward set of ok-to-stop functions for the remaining daemon types, however ha-rgw would require introduction of daemon types to ok-to-stop (currently the classes self.TYPE, which is a service type, and causes issues for ha-rgw). Since it's a more substantial change that ultimately affects all ok-to-stop functions by changing parameters we're going to split it into its own PR

**Comment by Daniel-Pivonka** [comment]
> This appears to duplicate a fair amount of logic for if the num of dameons >=1. Can it be consolidated up into the existing `ok_to_stop` of the CephadmService class?

i have something for this but it involves modifying the _enough_daemons_to_stop func so it will be easier to add once #38854 merged

File: src/pybind/mgr/cephadm/inventory.py [hunk 2]
```diff
@@ -9,6 +9,7 @@
 from ceph.deployment.service_spec import ServiceSpec
 from ceph.utils import str_to_datetime, datetime_to_str, datetime_now
 from orchestrator import OrchestratorError, HostSpec, OrchestratorEvent
+from orchestrator._interface import daemon_type_to_service, service_to_daemon_types
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/inventory.py:12]: please don't import from `orchestrator._interface`. instead please import them from `orchestrator` directly.

File: src/pybind/mgr/cephadm/inventory.py [hunk 3]
```diff
@@ -459,7 +460,7 @@ def get_daemons_by_type(self, service_type):
         result = []   # type: List[orchestrator.DaemonDescription]
         for host, dm in self.daemons.items():
             for name, d in dm.items():
-                if d.daemon_type == service_type:
+                if d.daemon_type in service_to_daemon_types(service_type):
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/inventory.py:463]: can you split this into a different commit? this is not related to ok-to-stop

File: src/pybind/mgr/cephadm/services/iscsi.py [hunk 4]
```diff
@@ -120,3 +121,21 @@ def get_set_cmd_dicts(out: str) -> List[dict]:
...
+        warn, warn_message = self._enough_daemons_to_stop(self.TYPE, daemon_ids, 'Iscsi', 1, True)
+        if warn:
+            return HandleCommandResult(-errno.EBUSY, '', warn_message)
+
+        # if reached here, there is > 1 nfs daemon. make sure none are down
+        warn_message = ('ALERT: 1 iscsi daemon is already down. ' +
+                        'Please bring it back up before stopping this one')
+        iscsi_daemons = self.mgr.cache.get_daemons_by_type(self.TYPE)
+        for i in iscsi_daemons:
+            if i.status != 1:
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/services/iscsi.py:136]: would be great to have an `Enum` here. Maybe like so:

https://github.com/ceph/ceph/blob/39fd806cd45e991deac49c8a938dd6381306facf/src/pybind/mgr/orchestrator/module.py#L33-L37

do you want to add this to the `_interface.py`?

**Review by sebastian-philipp — APPROVED** [review]

**Comment by sebastian-philipp** [comment]
ping @jmolmo

**Review by jmolmo — APPROVED** [review]
LGTM

File: src/pybind/mgr/orchestrator/_interface.py [hunk 5]
```diff
@@ -1456,6 +1457,12 @@ def yaml_representer(dumper: 'yaml.SafeDumper', data: 'DaemonDescription') -> An
         return dumper.represent_dict(data.to_json().items())
 
 
+class DaemonDescriptionStatus(enum.IntEnum):
+    error = -1
+    stopped = 0
+    running = 1
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator/_interface.py:1463]: https://github.com/ceph/ceph/blob/05ef9e1ad99f5d8d8bf75a02103da5d98d7e1da2/qa/tasks/cephadm_cases/test_cli.py#L36

```yaml
daemon_type: osd
daemon_id: '2'
hostname: smithi142
container_id: a93bbc870330
container_image_id: 2f759791c60922712d18300934155125b3d685a60a411248cbe486ce4489bd7a
container_image_name: quay.ceph.io/ceph-ci/ceph:52fc503cf18cf3bb446b840ba00be073017b8373
container_image_digests:
- quay.ceph.io/ceph-ci/ceph@sha256:42659bfb76ef5f335b56e9a900612c30b98b66ea577069fbc2d31c94a9b434c6
version: 17.0.0-699-g52fc503c
status: !!python/object/apply:orchestrator._interface.DaemonDescriptionStatus
- 1
status_desc: running
osdspec_affinity: None
is_active: false
last_refresh: '2021-02-12T16:38:43.947156Z'
created: '2021-02-12T16:30:13.999057Z'
started: '2021-02-12T16:30:13.739594Z'
```
```
sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph log 'Ended test tasks.cephadm_cases.test_cli.TestCephadmCLI.test_yaml'
test_yaml (tasks.cephadm_cases.test_cli.TestCephadmCLI) ... FAIL

======================================================================
FAIL: test_yaml (tasks.cephadm_cases.test_cli.TestCephadmCLI)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_ceph-c_52fc503cf18cf3bb446b840ba00be073017b8373/qa/tasks/cephadm_cases/test_cli.py", line 36, in test_yaml
    self.assertNotIn('!!python', out)
``` 

https://pulpito.ceph.com/swagner-2021-02-11_11:00:52-rados:cephadm-wip-swagner3-testing-2021-02-10-1322-distro-basic-smithi/5874628

I **never** thought this dead simple test would save us 8 months later

**Review by sebastian-philipp — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/orchestrator/_interface.py (hunk 5 above)
Comment (Daniel-Pivonka) [reply src/pybind/mgr/orchestrator/_interface.py]: i removed the DaemonDescriptionStatus changes as i could not get this to work and would like to get this merged and that was not really related to the main goal of this pr. changes are here for future reference https://github.com/Daniel-Pivonka/ceph/commit/bd92843d0a59cf2e90227e1e39b5cabef7dcdbf3

**Comment by sebastian-philipp** [comment]
please add either iscsi or NFS to https://github.com/ceph/ceph/blob/a171b32a2110bbb548b242875238d6c65a6e0bbb/qa/suites/rados/cephadm/upgrade/fixed-2.yaml in order to verify this in teuthology.

**Comment by sebastian-philipp** [comment]
https://pulpito.ceph.com/swagner-2021-03-02_10:27:34-rados:cephadm-wip-swagner-testing-2021-03-01-1657-distro-basic-smithi/

**Review by sebastian-philipp — APPROVED** [review]
